src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+  #            sing[ising].restype.g, sing[ising].restype.k,
+  #            sing[ising].restype.eta, sing[ising].restype.rho,
+  #            sing[ising].restype.taua, sing[ising].restype.taur)
+    end
+
+    # Find next ising
+    while true
+        ising += 1
+        if ising > intr.msing || intr.psilim < intr.sing[min(ising, intr.msing)].psifac
+            break
+        end
+        q = intr.sing[ising].q
+        if intr.mlow <= ctrl.nn * q && intr.mhigh >= ctrl.nn * q
+            break
+        end
+    end
+
+    # Compute conditions at next singular surface
+    if ising > intr.msing || intr.psilim < intr.sing[min(ising, intr.msing)].psifac
+        psimax = intr.psilim * (1 - eps)
+        m1 = round(Int, ctrl.nn * intr.qlim) + sign(one, ctrl.nn * equil.sq.fs1[mpsi, 4])
+        next = "finish"
+    else
+        psimax = sing[ising].psifac - ctrl.singfac_min / abs(ctrl.nn * intr.sing[ising].q1)
+        m1 = round(Int,ctrl.nn * intr.sing[ising].q)
+        println(crit_out_unit, "ising = $ising, psifac = $(intr.sing[ising].psifac), q = $(intr.sing[ising].q), di = $(intr.sing[ising].di), alpha = $(intr.sing[ising].alpha)")
+    end
+
+    # Restart ode solver
+    istate = 1 # is this in use?
+    odet.istep += 1
+    rwork[1] = psimax
+    new = true
+    u_save .= u
+    psi_save = odet.psifac
+
+    # Write to files
+    println(crit_out_unit, "step info")
+    if crit_break
+        write(crit_bin_unit)
+    end
+
+end
+
+# Example stub for kinetic crossing
+function ode_kin_cross()
+    # Implement kinetic crossing logic here
+    return
+end
+
+# Example stub for resistive crossing
+function ode_resist_cross()
+    # Implement resistive crossing logic here
+    return
+end
+
+function ode_step(ising::Int, odet::OdeState, equil::Equilibrium.PlasmaEquilibrium, intr::DconInternal, ctrl::DconControl)
+    # ODE integrator step
+    # Compute relative tolerances
+    singfac_local = typemax(Float64)
+    if ctrl.kin_flag
+ #       if ising == 1 && intr.kmsing >= 1
+ #           singfac_local = abs(odet.psifac - kinsing[ising].psifac) /
+ #                           (kinsing[ising].psifac - psilow)
+ #       elseif ising <= intr.kmsing
+ #           singfac_local = min(
+ #               abs(odet.psifac - kinsing[ising].psifac),
+ #               abs(odet.psifac - kinsing[ising-1].psifac)
+ #           ) / abs(kinsing[ising].psifac - kinsing[ising-1].psifac)
+ #       end
+    else
+        if ising <= intr.msing
+            singfac_local = abs(intr.sing[ising].m - ctrl.nn * odet.q)
+        end
+        if ising > 1
+            singfac_local = min(
+                singfac_local,
+                abs(intr.sing[ising-1].m - ctrl.nn * q)
+            )
+        end
+    end
+    tol = singfac_local < ctrl.crossover ? ctrl.tol_r : ctrl.tol_nr
+    rtol = tol
+
+    # Compute absolute tolerances
+    for ieq in 1:2, isol in 1:odet.msol
+        atol0 = maximum(abs.(u[:, isol, ieq])) * tol
```

Comment:
Two "errors":

1. You have the slicing issue, so want [`@view` or `@views`](https://docs.julialang.org/en/v1/base/arrays/#Views-(SubArrays-and-other-view-types))
2. `abs.()` will allocate an array.

You can do the following to avoid allocations.
`@views atol0 = maximum(abs, u[:, isol, ieq]) * tol`